### ISSUE TITLE
[tune] Appending to a list after each trial

### ISSUE THREAD

**Issue by athenawisdoms** [ray-project/ray#8052]
# What is your question?

Is it possible to append some data to a list after running each trial?

At the end of each trial, I will like to calculate some custom evaluation metric, then store them in a common list `results`, which I can then retrieve after all the trials are done.

The following code is an attempt to do so, but after all the trials are done, `results` list is still empty.

```
import torch.optim as optim
from ray import tune
from ray.tune.examples.mnist_pytorch import get_data_loaders, ConvNet, train, test

results = []                                    # <===== trying to store some data here

def eval(model, test_loader):
    """ Dummy custom evaluation metric calculation """
    return 1 / test(model, test_loader)

def train_mnist(config):
    global result

    train_loader, test_loader = get_data_loaders()
    model = ConvNet()
    optimizer = optim.SGD(model.parameters(), lr=config["lr"])
    for i in range(10):
        train(model, optimizer, train_loader)
        acc = test(model, test_loader)
        tune.track.log(mean_accuracy=acc)

    m = eval(model, test_loader)
    print('Custom eval metric:', m)
    results.append(m)

analysis = tune.run(
    train_mnist,
    config={"lr": tune.grid_search([0.2, 0.1]) }
)

print(results)                                   # <===== returns an empty list []
```

*Ray version and other system information (Python version, TensorFlow version, OS):*

- ray 0.8.4
- torch 1.3.1
- python 3.7.3
- Ubuntu 18.04

**Comment by anabranch** [comment]
Hey Athena,

the problem is that the list is on your local machine. Since tune is distributed, all of these functions are actually going to run on a worker - so the global reference doesn't actually work.

If I'm not mistaken, you're probably going to need to create an actor to do this.

However, what's the end result of what you're trying to do? Why do you want to create this list in the first place?

**Comment by athenawisdoms** [comment]
@anabranch Hello thanks for the swift reply!

I like to calculate custom evaluation metrics after every trial, whose values are more relevant for my project than accuracy and loss. 

After all the trials are completed, I will like to collect all these custom metrics into a list/DataFrame which I will then use to select the best configs for each custom metric.

Is there an easier alternative to creating an actor?

**Comment by anabranch** [comment]
I think the following tutorial might be more helpful for you: https://gist.github.com/anabranch/417cdc19b7f855bcfc96e0bb1d14ec9e

you can't log whatever you want with the tune.track.log and it will give you the results per trial and you can sort by that value.

**Comment by anabranch** [comment]
Let me know if that doesn't clarify it for you.

**Comment by anabranch** [comment]
ping @athenawisdoms !

**Comment by anabranch** [comment]
Closing this out, re-open if the above didn't solve your issue.
